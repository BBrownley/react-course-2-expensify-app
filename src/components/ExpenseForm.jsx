import React, { useState } from "react";

import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

import moment from "moment";

const ExpenseForm = props => {
  const [description, setDescription] = useState(
    props.expense ? props.expense.description : ""
  );
  const [amount, setAmount] = useState(
    props.expense ? (props.expense.amount / 100).toFixed(2) : ""
  );
  const [note, setNote] = useState(props.expense ? props.expense.note : "");
  const [date, setDate] = useState(
    props.expense ? moment(props.expense.createdAt) : moment()
  );
  const [dateFocused, toggleDateFocused] = useState(false);
  const [errors, setErrors] = useState([]);

  const onAmountChange = amount => {
    if (amount.match(/^\d*(\.\d{0,2})?$/)) {
      setAmount(amount);
    }
  };

  const onSubmit = e => {
    e.preventDefault();

    setErrors([
      ...(!description ? ["Please fill out the description field"] : []),
      ...(!amount ? ["Please fill out the amount field"] : [])
    ]);

    if (description && amount) {
      props.onSubmit({
        description,
        amount: parseFloat(amount) * 100,
        note,
        createdAt: date.valueOf() // in ms
      });
    }
  };

  return (
    <form onSubmit={e => onSubmit(e)}>
      <br />
      <br />
      <input
        type="text"
        placeholder="Description"
        autoFocus
        value={description}
        onChange={e => setDescription(e.target.value)}
      />
      <br />
      <br />
      <input
        type="text"
        placeholder="Amount"
        autoFocus
        value={amount}
        onChange={e => onAmountChange(e.target.value)}
      />
      <br />
      <br />
      <textarea
        placeholder="Add a note for your expense (optional)"
        value={note}
        onChange={e => setNote(e.target.value)}
      ></textarea>
      <br />
      <br />
      <SingleDatePicker
        date={date}
        onDateChange={newDate => {
          if (newDate) {
            setDate(newDate);
          }
        }}
        focused={dateFocused}
        onFocusChange={({ focused }) => toggleDateFocused(focused)}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
      <br />
      <br />
      <ul>
        {errors.map((error, index) => (
          <li className="expense-form__error" key={index}>
            {error}
          </li>
        ))}
      </ul>
      <button>{props.expense ? "Edit" : "Add"} expense</button>
    </form>
  );
};

export default ExpenseForm;
