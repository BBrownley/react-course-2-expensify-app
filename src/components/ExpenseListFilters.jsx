import React, { useState } from "react";
import { connect } from "react-redux";

import { DateRangePicker } from "react-dates";

import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters.js";
//random comment

export const ExpenseListFilters = props => {
  const [focusedInput, setFocusedInput] = useState(null);

  const onDatesChange = ({ startDate, endDate }) => {
    props.setStartDate(startDate);
    props.setEndDate(endDate);
  };

  const onTextChange = e => props.setTextFilter(e.target.value);

  const onSortByChange = e => {
    if (e.target.value === "date") {
      props.sortByDate();
    } else if (e.target.value === "amount") {
      props.sortByAmount();
    }
  };

  return (
    <div>
      <input type="text" value={props.filters.text} onChange={onTextChange} />
      <select value={props.filters.sortBy} onChange={onSortByChange}>
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
      <DateRangePicker
        startDate={props.filters.startDate}
        startDateId={"Fgsfdgss"}
        endDate={props.filters.endDate}
        endDateId={"fshfgushf"}
        onDatesChange={onDatesChange}
        focusedInput={focusedInput}
        onFocusChange={focusedInput => setFocusedInput(focusedInput)}
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
    </div>
  );
};

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => {
  return {
    setStartDate: startDate => dispatch(setStartDate(startDate)),
    setEndDate: endDate => dispatch(setEndDate(endDate)),
    setTextFilter: value => dispatch(setTextFilter(value)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
