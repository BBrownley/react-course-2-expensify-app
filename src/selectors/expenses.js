import moment from "moment";

const getVisibleExpenses = (expenses, filters) => {
  const { text, sortBy, startDate, endDate } = filters;

  console.log(startDate, expenses);

  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);

      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment)
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment)
        : true;

      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
