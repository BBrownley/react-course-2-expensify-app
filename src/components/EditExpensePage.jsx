import React from "react";
import { connect } from "react-redux";

import { editExpense, removeExpense } from "../actions/expenses";

import ExpenseForm from "./ExpenseForm";

export const EditExpensePage = props => {
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={expense => {
          props.editExpense(props.expense.id, expense);
          props.history.push("/");
        }}
      />
      <button
        onClick={() => {
          props.removeExpense(props.expense.id);
          props.history.push("/");
        }}
      >
        Remove Item
      </button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = dispatch => {
  return {
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: id => dispatch(removeExpense({ id }))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
