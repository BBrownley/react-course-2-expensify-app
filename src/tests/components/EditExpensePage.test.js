import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let wrapper, editExpense, removeExpense, history;

describe("EditExpensePage", () => {
  beforeEach(() => {
    editExpense = removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
      <EditExpensePage
        editExpense={editExpense}
        removeExpense={removeExpense}
        history={history}
        expense={expenses[2]}
      />
    );
  });

  it("should render EditExpensePage", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle editExpense", () => {
    wrapper.find("ExpenseForm").prop("onSubmit")({
      description: "Soda",
      amount: 2000
    });
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, {
      description: "Soda",
      amount: 2000
    });
    expect(history.push).toHaveBeenLastCalledWith("/");
  });

  it("should handle removeExpense", () => {
    wrapper.find("button").simulate("click");
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[2].id);
    expect(history.push).toHaveBeenLastCalledWith("/");
  });
});
