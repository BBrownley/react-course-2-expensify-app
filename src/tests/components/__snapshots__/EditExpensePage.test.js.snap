// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePage should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 1000,
        "createdAt": "1969-12-28T00:00:00.000Z",
        "description": "Water",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Item
  </button>
</div>
`;
